CIPs/0020.md

```diff
@@ -0,0 +1,127 @@
+# CIP [0020]: Generic hash function support via extensible precompile
+
+- Date: 2020-09-4
+- Author: @prestwich (James Prestwich)
+- Status: WIP
+
+### Terminology
+
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as described in 
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html).
+
+## Overview
+
+As a descendant of the EVM, Celo supports only a handful of common hash 
+functions. This restricts the design-space of Celo protocols, and 
+
+## Goals
+
+- Provide support for common cryptographic hash functions
+- Expand options for on-chain verification of protocols not specifically 
+designed for EVM compatibility
+    - e.g. Verification of off-chain signatures using Blake2Xs, including Celo's 
+    own epoch commitments within a smart contract.
+    - e.g. Verification of remote Proof of Work from a wide variety of chains
+- Provide a clear route for acceptance and adoption of additional hash functions
+
+## Proposed Solution
+
+Introduce a new precompiled contract at address [TBD] that provides access to 
+a set of pre-determined hash functions. The input to this pre-compile consists 
+of a one-byte selector, a function-specific parameter block, and the preimage. 
+The output of this pre-compile is dependent on the selector, and may be fixed- 
+or variable-length. If the selector is unknown (not on the current 
+compatibility list), or the function-specific, the precompile MUST return an 
+error.
```

Comment:
Does `return an error` here mean an error indicator value, or does it need to revert the transaction?

Reply:
it means the same as in EIP-2537. The output of the EVM staticcall should be `(false, revertData)`